### ISSUE TITLE
[autoscaler] Issues with lack of parallelisation (larger cluster sizes)

### ISSUE THREAD

**Issue by ls-daniel** [ray-project/ray#3599]
### System information
- Ubuntu 18.04
- Ray installed from source (git master)
- Python 3.6
- `ray up` with a yaml file that has about 100 or more ec2 nodes

### Describe the problem
The Ray autoscaler is designed in such a way that there are various issues with starting a lot of nodes quickly (for example, if you set min_workers=max_workers=200).

I expect that 50 nodes should take approximately the same amount of time to start as 500. In practice what happens is that beyond about 100 or so things start to become impractically slow and _update doesn't run very often until everything is up fully.

One issue is that the 'get_node_ip' call in NodeUpdater is run synchronously in __init__, which can make the StandardAutoscaler _update loop take a long time (e.g. if this call takes 200ms per node, that's 20 seconds for 100 nodes best case). It's not run under the NodeUpdaterProcess (because the init call isn't) so it's serial.

Another problem is that this call (and possibly other API calls) aren't batched; so what ends up happening if you try to parallelise that, is that you hit EC2 rate limiting and get_node_ip can take tens of seconds to return.

### Source code / logs
This is an excerpt from a longer log of a slightly modified Autoscaler which has 'set_ssh_ip' split into a method on NodeUpdater.

```

Dec 21 15:50:40 Dec 21 15:50:40 NU set_ssh_ip done in 122.53ms
Dec 21 15:50:43 Dec 21 15:50:43 NU set_ssh_ip done in 83.736ms
Dec 21 15:50:43 Dec 21 15:50:43 NU set_ssh_ip done in 36102.643000000004ms
Dec 21 15:50:48 Dec 21 15:50:48 NU set_ssh_ip done in 166.25ms
Dec 21 15:50:49 Dec 21 15:50:49 NU set_ssh_ip done in 153.125ms
Dec 21 15:50:52 Dec 21 15:50:52 NU set_ssh_ip done in 131.434ms
Dec 21 15:50:53 Dec 21 15:50:53 NU set_ssh_ip done in 102.10900000000001ms
Dec 21 15:50:56 Dec 21 15:50:56 NU set_ssh_ip done in 41360.066000000006ms
Dec 21 15:50:57 Dec 21 15:50:57 NU set_ssh_ip done in 37164.039000000004ms
Dec 21 15:51:11 Dec 21 15:51:11 NU set_ssh_ip done in 116.062ms
Dec 21 15:51:12 Dec 21 15:51:12 NU set_ssh_ip done in 220.691ms
Dec 21 15:51:12 Dec 21 15:51:12 NU set_ssh_ip done in 198.95399999999998ms
Dec 21 15:51:12 Dec 21 15:51:12 NU set_ssh_ip done in 195.766ms
Dec 21 15:51:13 Dec 21 15:51:13 NU set_ssh_ip done in 203.863ms
Dec 21 15:51:15 Dec 21 15:51:15 NU set_ssh_ip done in 148.352ms
Dec 21 15:51:15 Dec 21 15:51:15 NU set_ssh_ip done in 288.272ms
Dec 21 15:51:27 Dec 21 15:51:27 NU set_ssh_ip done in 80317.613ms
Dec 21 15:51:44 Dec 21 15:51:44 NU set_ssh_ip done in 357.942ms
Dec 21 15:51:46 Dec 21 15:51:46 NU set_ssh_ip done in 213.292ms
Dec 21 15:51:58 Dec 21 15:51:58 NU set_ssh_ip done in 245.304ms
Dec 21 15:52:03 Dec 21 15:52:03 NU set_ssh_ip done in 371.163ms

```

**Comment by ls-daniel** [comment]
For clarity, NodeLauncher seems fine (I haven't looked too deeply but I believe this to use batch calls)

**Comment by ericl** [comment]
Yeah, I think this is an area that could use a lot more optimization.

Though one note: 500 nodes is over 32k cores! I wouldn't be surprised if there were other scheduler bottlenecks discovered at this scale, since we don't test releases at more than a couple hundred nodes I believe.

**Comment by richardliaw** [comment]
@ls-daniel thanks for opening this (and taking the time to read through the code)! 

Did you manage to come up with a workaround or have any thoughts? 

I suspect a fix for this immediate problem is to introduce some batch call for `self.ec2.instances.filter` that alleviates this bottleneck, although it's possible that at 100 nodes you'll hit the rate limiter pretty often.

**Comment by ls-daniel** [comment]
@ericl For my use case the autoscaler seems to be the major issue here. Though I don't doubt that eventually you'd run into issues there too.

@richardliaw right. I'm only looking at EC2 here though I suspect the same would apply to other providers.

Ignoring setup/teardown stuff, some general mumbling I expect to be useful for anyone that looks at this (I'll continue to investigate over the coming days...)

## AWSNodeProvider.nodes()

NodeLauncher calls indirectly. Launching 200 nodes with a batch size of 25 will hit ~16 times over the interval. I don't think this matters (let's say we have a rate limit of 5/s then it takes 3 seconds, fine).

StandardAutoscaler's _update method calls indirectly via self.workers() between 1 and 5(?) times depending on actions taken. During ramp up I think it should only call self.workers() three times at maximum (initially, after completed updates, and implicitly in launch_new_nodes). Provided we don't update too frequently (default is to update once every 5 seconds) I don't think this matters.

In summary I don't think this matters.

## AWSNodeProvider._node()

This call is unbatched: `self.ec2.instances.filter(InstanceIds=[node_id])`

True to name, it's an internal method. Called from:
* `AWSNodeProvider.is_running()`
Called from updater (could be cached)
Called from autoscaler (can_update; from recover_if_needed and update_if_needed)
* `AWSNodeProvider.is_terminated()`
Called from updater (set_ssh_ip, do_update; whilst checking for SSH access)
* `AWSNodeProvider.node_tags()`
autoscaler launch_config_ok (can_update)
autoscaler files_up_to_date (can_update)
* `AWSNodeProvider.external_ip()` (cached)
updater `get_node_ip`
* `AWSNodeProvider.internal_ip()` (cached)
updater `get_node_ip`
autoscaler _update (multiple calls here; list comprehension; recover_if_needed)
(this would break _horribly_ without caching)
* `AWSNodeProvider.set_node_tags()`
all in the updater.
after a successful update
* `AWSNodeProvider.terminate_node()`
don't care

**Comment by ls-daniel** [comment]
Spot on. I've fiddled with this a bit today. By using NodeUpdaterThread and by tweaking a little elsewhere (internal cache for the AWS Tags rather than calling out) I've been able to get significant speedups (200 nodes starting in approx 2 minutes from scratch, including starting the head node).

Tinkering continues.

**Comment by ericl** [comment]
Nice! I guess the use of processes wasn't the best decision before. One
trade-off is that it is harder to cancel pending requests since you can't
easily kill threads, though I don't think we take advantage of this anyways
at the moment.

**Comment by ls-daniel** [comment]
More progress with `min_workers = max_workers = 100`. The logs here are from a custom log formatting script I've written for debugging.

```
    1188ms | 15:46:43,226 - sherlock - [Timer   1188ms] AWSNodeProvider: ec2 call create_instances 1
... some time passes; approximately 45 seconds from `ray up` to the head node running ...
           | 15:47:29,632 - ip-10-2-246-53 - StandardAutoscaler: 0/100 target nodes
... some time passes; approximately 50 seconds until 100 NumNodesConnected in LoadMetrics ...
           | 15:48:20,517 - ip-10-2-246-53 - StandardAutoscaler: 100/100 target nodes
```

So now, for N=100 at least, it's as expected; the head and worker nodes take about the same amount of time.

Problem is, in order to achieve this, I've had to knock out the AWS create_tags calls because they seem to a) get called synchronously in-line; but more crucially b) get rate limited and cause all manner of slowness eventually (e.g. 20 seconds required for a call).

I've knocked up a quick HTTP server (no extra deps, just Python builtin stuff) that displays node status anyway; it runs as a thread in autoscaler.py and just dumps JSON with some internal stats when you visit it.

As for the tags; they are still necessary for operations that update configuration. I'd like to replace them with another system. It's mainly the 'launch config' and 'runtime config' hashes that are problematic. They could either be saved as hashes on the nodes themselves, or (my preference) just dumped to disk on the head node. I don't really mind, but I do think that EC2 limits make using instance tags too slow for this purpose (as far as I can tell you can't batch it).

**Comment by ls-daniel** [comment]
Indeed. At the moment I am using in-memory tracking.
The old AWSNodeProvider 'tags' calls just become getters/setters of that dict.

WRT persistence of tag state; I think the best thing to do would be to just write them to disk on the worker nodes. I'll have a tinker about today and test some stuff out. :)

**Comment by ls-daniel** [comment]
For clarity: the issues are with TAG_RAY_NODE_STATUS and TAG_RAY_RUNTIME_CONFIG, which are updated on the fly in unbatched calls.

The other four tags (NODE_NAME, NODE_TYPE, LAUNCH_CONFIG, CLUSTER_NAME) are fine because they're set as part of the create call.

So those four can be left as is, and the other two can live on a node in a ~/ray_status.json file or similar.

**Comment by ls-daniel** [comment]
Other excitements: ssh can be sped up massively by keeping a persistent connection.

In my testing it takes the time for a minimal ssh command (e.g. `ls`) to run from ~700ms to ~25ms. With twenty setup command that's 10s knocked off every node startup.

**Comment by richardliaw** [comment]
Oh that's a big improvement! 

Just wondering, will you be opening up a PR to push some changes upstream?

**Comment by ls-daniel** [comment]
Of course! Sorry, I'm still hacking about on this, just thought I would update with my findings.

**Comment by richardliaw** [comment]
Yeah, not a problem; just wanted to know! Thanks for opening that PR.

**Comment by ls-daniel** [comment]
I've updated the PR, rebased on master, this is running well here for me.

EC2: Initial node setup takes ~45 seconds; starting 85 nodes (with initial_workers, or with min_workers=max_workers set to 85) takes another 45s-1min.

Autoscaling also works much better because it's more responsive in general; there are fewer stalls (autoscaler._update() can still take longer than five seconds; I considered making this faster YAGNI)

**Comment by richardliaw** [comment]
Closed via #3720 I think; feel free to reopen!
